pymtl3/dsl/test/set_param_test.py

```diff
@@ -0,0 +1,178 @@
+  assert Z.houses[1].animals[0].meats[0].lunch == "bamboo"
+  assert Z.houses[1].animals[0].meats[1].lunch == "grass"
+
+  assert isinstance( Z.houses[2].animals[0], Dromaius )
+  assert isinstance( Z.houses[2].animals[1], HoneyBadger )
+  # FIXME
+  assert Z.houses[2].animals[0].lunch  == "bugs"
+  assert Z.houses[2].animals[0].dinner == "bugs"
+  assert Z.houses[2].animals[1].lunch  == "grass"
+  assert Z.houses[2].animals[1].dinner == "dirt"
+
+def test_multi_regex():
+  Z = Zoo()
+  Z.set_param( "top.construct", AnimalHouseTypes=[ PandaHouse, TigerTerrace, AnimalHouse, Aquarium ] )
+  Z.set_param( "top.houses[0].construct", AnimalTypes=[ Panda, Panda, Panda   ] )
+  Z.set_param( "top.houses[1].construct", AnimalTypes=[ Tiger, BabyTiger      ] )
+  Z.set_param( "top.houses[2].construct", AnimalTypes=[ Dromaius, HoneyBadger ] )
+  Z.set_param( "top.houses[0].animals[0].construct", lunch="potato", dinner="bamboo" )  
+  Z.set_param( "top.houses*.animals*.construct", lunch="onion" )
+  Z.set_param( "top.houses*.animals[0].construct", lunch="potato" )  
+  Z.elaborate()
+
+  assert Z.houses[0].animals[0].lunch  == "onion"
+  assert Z.houses[0].animals[0].dinner == "bamboo"
+
+def test_set_param_hierarchical():
+  Z = Zoo()
+
+  Z.set_param( "top.construct", AnimalHouseTypes=[ Aquarium, TigerTerrace ] )
+  Z.set_param( "top.houses[1].construct", AnimalTypes=[ Tiger, Panda ] )
+  Z.set_param( "top.houses*.animals*.construct", lunch="grass" )
+  Z.set_param( "top.houses*.animals[1].construct", lunch="meat" )
+
+  #TODO: Whether we should check throw an error in following scenario
+  Z.set_param( "top.houses*.xxx*.construct", lunch="grass" )
+
+  Z.elaborate()
+
+  assert isinstance(Z.houses[1], TigerTerrace)
+  assert Z.houses[1].animals[0].lunch == "grass"
+  assert Z.houses[1].animals[1].lunch == "meat"
+
+def test_component():
+  class Incr( Component ):
+    def construct( s, DataType=Bits4(), incr_value=1 ):
+      s.in_ = InPort ( DataType )
+      s.out = OutPort( DataType )
+
+      s.incr_value = DataType( incr_value )
+
+      @s.update
+      def up_incr():
+        s.out = s.in_ + s.incr_value
+  
+  class IncrArray( Component ):
+    def construct( s, num_incrs=1, DataType=Bits4() ):
+      s.in_ = [ InPort ( DataType ) for _ in range( num_incrs ) ]
+      s.out = [ OutPort( DataType ) for _ in range( num_incrs ) ]
+
+      s.incrs = [ Incr( DataType=DataType ) for _ in range( num_incrs ) ]
+
+      for i in range( num_incrs ):
+        s.connect( s.in_[i], s.incrs[i].in_ )
+        s.connect( s.out[i], s.incrs[i].out )
+
+  A = Incr()
+  A.set_param( "top.construct", DataType=Bits8, incr_value=2 )
+  A.apply( SimpleSim )
+  A.in_ = 1
+  A.tick()
+  print( "A.out==", A.out )
+  assert A.out == 3
+
+  B = IncrArray()
+  B.set_param( "top.construct", DataType=Bits8, num_incrs=3 )
+  B.set_param( "top.incrs*.construct", incr_value=3 )
+  B.set_param( "top.incrs[2].construct", incr_value=5 )
+  B.apply( SimpleSim )
```

Comment:
Can you use simple_sim_pass here? We don't want to import stuff from passes/ in dsl.